packages/form/src/SelectList.js

```diff
@@ -65,29 +74,48 @@ class SelectList extends PureComponent {
         multiple: false,
         showCheckAll: true,
         minCheck: 0,
-        allOptionLabel: 'All',
-        values: undefined,
-        defaultValues: [],
+        checkAllLabel: 'All',
+        value: undefined,
+        defaultValue: undefined,
         onChange: () => {},
         title: undefined,
         desc: undefined,
     };
 
     state = {
-        // eslint-disable-next-line react/destructuring-assignment
-        checkedState: getInitialCheckedState(this.props.values || this.props.defaultValues),
+        checkedState: getInitialCheckedState(
+            this.getInitialValue(),
+            this.props.multiple
+        ),
     };
 
     componentWillReceiveProps(nextProps) {
+        warning(
+            this.getIsControlled(this.props) === this.getIsControlled(nextProps),
+            '<SelectList> should not switch from controlled to uncontrolled (or vice versa).'
+        );
+
         if (this.getIsControlled(nextProps)) {
             this.setState({
-                checkedState: getInitialCheckedState(nextProps.values),
+                checkedState: getInitialCheckedState(nextProps.value, nextProps.multiple),
+            });
+        } else if (this.props.multiple !== nextProps.multiple) {
+            warning(false, '<SelectList>: you should not change `multiple` prop while it is uncontrolled. Its value will be reset now.');
+            this.setState({
+                checkedState: getInitialCheckedState([])
             });
         }
     }
 
+    getInitialValue() {
+        const { value, defaultValue, multiple } = this.props;
+        const newDefaultValue = (defaultValue === undefined && multiple) ? [] : defaultValue;
+
+        return (value !== undefined) ? value : newDefaultValue;
+    }
```

Comment:
Instead of multiple ternaries, I would suggest simple `if` blocks to make the logic look clearer.

```js
getInitialValue() {
    const { value, defaultValue, multiple } = this.props;

    if (value !== undefined) {
        return value;
    }
    
    if (multiple && defaultValue === undefined) {
        return [];
    }

    return defaultValue;
}
```

And since the logic is almost the same as `<SelectRow>`, maybe it's a good time to extract it to a helper function?

Reply:
Good refactor!
I think it's still ok that we write this function twice, it is not a too complicated logic.